### PR TITLE
Revert futures 0.2 changes

### PR DESCRIPTION
**Pull request by seanmonstar**
See #1481 

:cry:

### CODE REVIEW

**Comment by srijs** [comment]
💔

From the CI errors, it looks like we might need to back-port some of the fixes I included for the new tokio behaviour. I'll see if I can figure out which ones they are!

**Comment by kamyuentse** [comment]
On my local machine, the test `conn_drop_prevents_pool_checkout` can pass some time.
Seems there is a race somewhere, I am trying locate the problem.

**Comment by srijs** [comment]
Yeah, it‘s a race condition I fixed as part of the futures 0.2 changes
that are being reverted here, should be a matter of backporting it!

File: tests/client.rs [hunk 1]
```diff
@@ -1043,27 +1042,28 @@ mod dispatch_impl {
             let mut buf = [0; 4096];
             sock.read(&mut buf).expect("read 1");
             sock.write_all(b"HTTP/1.1 200 OK\r\nContent-Length: 0\r\n\r\n").unwrap();
-            //sock.read(&mut buf).expect("read 2");
+            sock.read(&mut buf).expect("read 2");
```
Comment (seanmonstar) [inline tests/client.rs]: Oh, this is what made it pass always :D

**Comment by kamyuentse** [comment]
I want to know why removing that test case.

**Comment by seanmonstar** [comment]
@kamyuentse I kept trying variants, and couldn't seem to get it passing 100% of the time without disabling the exact thing it was testing. And so then I read through it, trying to remember what the heck it was even testing, and realized it was because before inside `client/mod.rs` there was a `tx.start_send(req).unwrap()`. That doesn't exist anymore, and the test is really contrived.

Actually, I'm *fairly* certain the problem is the race condition in futures mpsc queue: https://github.com/rust-lang-nursery/futures-rs/issues/909

That'd make sense, since the test was basically dropping the `Receiver` in a runtime thread, and trying to use the `Sender` in the test thread, and occasionally seeing deadlocks as the message was never dropped.

**Comment by kamyuentse** [comment]
@seanmonstar Thanks for your explanation! I don't know whether futures issue cause tokio worker panic, but I think I should cc @carllerche .

**Pull request merged**